rotkehlchen/chain/evm/decoding/balancer/balancer_cache.py

```diff
@@ -0,0 +1,203 @@
+import logging
+from typing import TYPE_CHECKING, Final, Literal
+
+from rotkehlchen.assets.asset import UnderlyingToken
+from rotkehlchen.assets.utils import TokenEncounterInfo, get_or_create_evm_token
+from rotkehlchen.chain.evm.decoding.balancer.constants import (
+    CPT_BALANCER_V1,
+    CPT_BALANCER_V2,
+)
+from rotkehlchen.chain.evm.decoding.balancer.utils import (
+    query_balancer_pools,
+    query_balancer_pools_count,
+)
+from rotkehlchen.chain.evm.types import string_to_evm_address
+from rotkehlchen.constants.misc import ONE
+from rotkehlchen.errors.serialization import DeserializationError
+from rotkehlchen.fval import FVal
+from rotkehlchen.globaldb.cache import (
+    globaldb_get_general_cache_values,
+    globaldb_set_general_cache_values,
+    globaldb_update_cache_last_ts,
+)
+from rotkehlchen.globaldb.handler import GlobalDBHandler
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.serialization.deserialize import deserialize_evm_address, deserialize_int
+from rotkehlchen.types import CacheType, ChainID, ChecksumEvmAddress, EvmTokenKind
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.user_messages import MessagesAggregator
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+def query_balancer_data(
+        inquirer: 'EvmNodeInquirer',
+        cache_type: Literal[CacheType.BALANCER_V1_POOLS, CacheType.BALANCER_V2_POOLS],
+        msg_aggregator: 'MessagesAggregator',
+) -> tuple[set['ChecksumEvmAddress'], set['ChecksumEvmAddress']]:
+    """Query and store balancer pools with their gauges.
+
+    May raise:
+    - RemoteError
+    """
+    version: Final = 1 if cache_type == CacheType.BALANCER_V1_POOLS else 2
+    protocol = CPT_BALANCER_V1 if version == 1 else CPT_BALANCER_V2
+    latest_pools_count = query_balancer_pools_count(
+        chain=inquirer.chain_id,
+        version=version,
+    )
+    gauge_key_parts = (str(inquirer.chain_id.value), str(version))
+    pool_key_parts = (str(inquirer.chain_id.value),)
+    with GlobalDBHandler().conn.read_ctx() as cursor:
+        existing_pools = {
+            string_to_evm_address(address)
+            for address in globaldb_get_general_cache_values(
+                cursor=cursor,
+                key_parts=(cache_type, *pool_key_parts),
+            )
+        }
+        existing_gauges = {
+            string_to_evm_address(address)
+            for address in globaldb_get_general_cache_values(
+                cursor=cursor,
+                key_parts=(CacheType.BALANCER_GAUGES, *gauge_key_parts),
+            )
+        }
+
+    if latest_pools_count == len(existing_pools):
+        with GlobalDBHandler().conn.write_ctx() as write_cursor:
```

Comment:
wouldn't it make more sense to save the globaldb in a variable? You are initilizating it several times calling the constructor

Reply:
okay.